server/sonar-server/src/main/java/org/sonar/server/computation/step/DuplicationMeasuresStep.java

```diff
@@ -0,0 +1,168 @@
+      new SumFormula(DUPLICATED_FILES_KEY),
+      new DuplicationFormula()
+      );
+  }
+
+  @Override
+  public void execute() {
+    FormulaExecutorComponentVisitor.newBuilder(metricRepository, measureRepository)
+      .buildFor(formulas)
+      .visit(treeRootHolder.getRoot());
+  }
+
+  private class DuplicationFormula implements Formula<DuplicationCounter> {
+
+    private final Metric nclocMetric;
+    private final Metric linesMetric;
+    private final Metric commentLinesMetric;
+
+    public DuplicationFormula() {
+      this.nclocMetric = metricRepository.getByKey(NCLOC_KEY);
+      this.linesMetric = metricRepository.getByKey(LINES_KEY);
+      this.commentLinesMetric = metricRepository.getByKey(COMMENT_LINES_KEY);
+    }
+
+    @Override
+    public DuplicationCounter createNewCounter() {
+      return new DuplicationCounter();
+    }
+
+    @Override
+    public Optional<Measure> createMeasure(DuplicationCounter counter, CreateMeasureContext context) {
+      return createDuplicatedLinesMeasure(counter, context)
+        .or(createDuplicatedLinesDensityMeasure(counter, context));
+    }
+
+    private Optional<Measure> createDuplicatedLinesMeasure(DuplicationCounter counter, CreateMeasureContext context) {
+      Optional<Integer> duplicatedLines = counter.duplicatedLinesCounter.getValue();
+      if (context.getMetric().getKey().equals(DUPLICATED_LINES_KEY) && context.getComponent().getType().isHigherThan(FILE) && duplicatedLines.isPresent()) {
+        return Optional.of(Measure.newMeasureBuilder().create(duplicatedLines.get()));
+      }
+      return Optional.absent();
+    }
+
+    private Optional<Measure> createDuplicatedLinesDensityMeasure(DuplicationCounter counter, CreateMeasureContext context) {
+      Optional<Integer> duplicatedLines = counter.duplicatedLinesCounter.getValue();
+      if (context.getMetric().getKey().equals(DUPLICATED_LINES_DENSITY_KEY) && duplicatedLines.isPresent()) {
+        Optional<Integer> nbLines = getNbLinesFromLocOrNcloc(context);
+        if (nbLines.isPresent() && nbLines.get() > 0) {
+          return Optional.of(Measure.newMeasureBuilder().create(calculateDensity(duplicatedLines.get(), nbLines.get())));
+        }
+      }
+      return Optional.absent();
+    }
+
+    private Optional<Integer> getNbLinesFromLocOrNcloc(CreateMeasureContext context) {
+      Optional<Measure> lines = measureRepository.getRawMeasure(context.getComponent(), linesMetric);
+      if (lines.isPresent()) {
+        return Optional.of(lines.get().getIntValue());
+      }
+      Optional<Measure> nclocs = measureRepository.getRawMeasure(context.getComponent(), nclocMetric);
+      if (nclocs.isPresent()) {
+        Optional<Measure> commentLines = measureRepository.getRawMeasure(context.getComponent(), commentLinesMetric);
+        int nbLines = nclocs.get().getIntValue();
+        return Optional.of(commentLines.isPresent() ? nbLines + commentLines.get().getIntValue() : nbLines);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)